### PR TITLE
fix: update linux media folder permissions to $USER:www-data (#836)

### LINKED ISSUE #836: Linux Install Instructions for qatrack/media folder Permissions
From google groups chat: https://groups.google.com/d/msgid/qatrack/7d3433e0-1c6a-4c00-92a0-d84cc0b0a664n%40googlegroups.com?utm_medium=email&utm_source=footer

> for the sake of completeness: I came across another small problem today when I tried to upload a file attachment to a serivce event.
> Since the supervisor services (gunicorn and django-q2) are now run with the QATrack+ user rights and no longer with the www-data user I had to adjust the permissions to some qatrack/media subfolder. 
> sudo chmod g+w media/uploads/serviceevent
> sudo chmod g+w media/uploads/testlist
> These two were missing the group write permission. Interestingly the other folders were ok.
> 
> The [install instructions](https://docs.qatrackplus.com/en/releases-4.0/install/linux.html#setting-up-django-q) for a new installation also change the settings to user:group ownership of www-data:www-data and not to $USER:www-data.
> This seems to be a bit inconsistent.

### LINKED ISSUE #844: Bug: Ubuntu Create Cache Table instruction Wrong Order
From google groups: https://groups.google.com/g/qatrack/c/fcVBZm4I5WI/m/cDoosJK1AQAJ?utm_medium=email&utm_source=footer&pli=1

### Problem

`docs/install/linux.rst` runs `python manage.py loaddata fixtures/defaults/*/*` (line ~374) but doesn't run `createcachetable` until ~40 lines later. Loading the default fixtures fires `post_save` handlers that invalidate cached counts (`qatrack/qa/signals.py:302`, `qatrack/service_log/signals.py:52`, `qatrack/faults/signals.py:15`), and the default backend is `DatabaseCache` (`qatrack/settings.py:300`). So `loaddata` always fails on a fresh install that follows the Linux docs:

```
django.db.utils.ProgrammingError: (1146, "Table 'qatrackplus40.qatrack_cache_table' doesn't exist")
```

(sqlite reproduces it as `Could not load service_log.ServiceEventStatus(pk=1): no such table: qatrack_cache_table`)

The Windows docs and `deploy/docker/django/entrypoint.sh` already have the correct order; Linux does not. Same root cause as #621, which was closed with a workaround rather than a doc fix — it's still being hit on 4.0 (see the [[mailing list thread](https://groups.google.com/g/qatrack/c/fcVBZm4I5WI)](https://groups.google.com/g/qatrack/c/fcVBZm4I5WI)).

Two follow-on problems in the same thread:

1. **A missing cache table takes down the error pages too.** `qatrack/context_processors.py` calls `cache.delete()` at module import, so an unusable cache backend crashes template rendering — including the 500 page that would have reported the real error. Users see an unexplained internal server error.
2. **`accounts.0001_initial` crashes with `__fake__.DoesNotExist`.** Line 9 does a bare `User.objects.get(username="QATrack+ Internal")`. That user is created by `qa.0005`, so the migration fails if `qa` migrations were faked or the user was deleted.

### Workaround for affected users

```bash
python manage.py createcachetable
python manage.py loaddata fixtures/defaults/*/*
```

### Suggested fix

1. `docs/install/linux.rst`: move `createcachetable` to immediately after `migrate`, before `loaddata`; add `createcachetable` to `docs/install/linux_upgrade_from_3.rst`.
2. `qatrack/context_processors.py`: guard the import-time invalidation (log a pointer to `createcachetable`) and let cache reads/writes in `site()` fall back to computing values directly, so a misconfigured cache costs performance instead of the whole site.
3. `qatrack/accounts/migrations/0001_initial.py`: use `get_or_create` for the internal user instead of `get`.

### Environment

QATrack+ 4.0, Ubuntu 22.04, MySQL (also reproduced on sqlite).

### PR DESCRIPTION
**Pull request by nsmela**
## Summary

Update Linux installation instructions, upgrade documentation, backup restore guide, and Django system check hints so that `logs` and `qatrack/media` folders are owned by `$USER:www-data` instead of `www-data:www-data`.

In QATrack+ v4.0+, Gunicorn and Django-Q2 run under the QATrack+ user account (`$USER`) rather than `www-data` (which only runs Nginx). Setting ownership to `www-data:www-data` in the install instructions caused permission errors when uploading attachments to Service Events and Test Lists because Python's default umask stripped group-write permissions on newly created subdirectories. Setting folder ownership to `$USER:www-data` with `chmod 2775` on directories and `chmod 664` on files ensures the application user can always create and write files, while allowing Nginx to read media and static files.

## Related Issues

Closes #836

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature ( adds functionality, always assumed to be breaking until tested by maintainers)
- [ ] Breaking change (fix or addition that changes existing behaviour)
- [x] Documentation only

## Target Branch

develop

## AI Usage Disclosure

- [ ] No AI tools were used in this PR.
- [x] AI tools were used:
  - **Tools / where used:** Antigravity IDE for drafting documentation updates and unit tests.
  - [x] I've reviewed every AI-generated line as if I wrote it myself, and confirm: no PHI shared, no license-incompatible content (Apache-2.0).

---

### Coder Tasks

- [x] Rebased / merged `develop` branch; no merge conflicts.
- [x] `uv run pre-commit run --all-files` passes (ruff lint, ruff-format, django-upgrade).
- [x] `uv run python -Wd manage.py check` reviewed for new deprecation warnings.
- [x] `uv run pytest` full test suite passes.
- [x] `uv run python manage.py makemigrations --check --dry-run` if migrations are found, change should be marked Breaking
- [x] Docker builds succeed if touched.
- [x] Docs updated under `docs/` if behaviour, settings, or APIs changed.
- [x] `release_notes.md` updated if user-facing.
- [x] No secrets, debug code, or stray `print`/`console.log` left in.

### Review Code

- [ ] Reviewer verifies the target branch.
- [ ] For non-trivial changes: pulled the branch, ran it locally, and confirmed it works as described.
- [ ] CI is green; test suite and `pre-commit` pass.
- [ ] Reads through all changed files for anything weird or unintended.
- [ ] Logic is correct; edge cases and error handling considered.
- [ ] Migrations are sane and reversible where possible.
- [ ] Changes are consistent with the modernization direction (no reintroduction of patterns being migrated away from, e.g. flake8/yapf-era conventions).
- [ ] No security concerns (input validation, permissions, injection, XSS).
- [ ] Comments are sufficient and not overkill.
- [ ] Typos, spellcheck, professional.
- [ ] Reviewer sends back review.

### Edits

- [ ] Coder fixes any issues.
- [ ] Reviewer approves.

---

## Testing Notes

- Unit tests added in `qatrack/qatrack_core/tests/test_checks.py` covering all branches of `check_media_folder_permissions` (all 6 tests passing).
- Validated with `manage.py check`.
- Verified Sphinx documentation build (`sphinx-build -b html docs docs/_build/html`) builds cleanly without warnings on modified `.rst` files.
- Verified `ruff check` and `ruff format`.

## Screenshots

N/A

## Errors

```FAILED qatrack/qa/tests/test_selenium.py::TestPerformQC::test_load_autosave - AssertionError: assert '1980-11-30 00:00' == '12 May 1980 12:00' ```

fixed in #842

### CODE REVIEW

**Comment by nsmela** [comment]
I haven't verified the instructions yet. I'm hoping to have more time in a few weeks.

**Comment by nsmela** [comment]
## Review: verified on a clean Ubuntu 24.04 VM

I ran `docs/install/linux.rst` end to end on a fresh Ubuntu 24.04 LTS VM (new non-root `qatrack` user with sudo, PostgreSQL 16, nginx, supervisor, gunicorn, django-q2), reproduced #836 against the pre-PR instructions, and then validated this PR's block against it.

### The fix is correct

**#836 reproduced first.** With the pre-PR ownership (`sudo chown -R www-data:www-data logs qatrack/media` + `chmod ug+rwxs`), an attachment write driven by the app user fails exactly as reported:

```
PermissionError: [Errno 13] Permission denied:
  '/home/qatrack/web/qatrackplus/qatrack/media/uploads/tmp'
```

**After this PR's block, it works.** The attachment lands in `media/uploads/testlist/1/…` and nginx serves it over HTTP 200:

```
drwxrwsr-x qatrack:www-data  qatrack/media           (2775)
drwxr-sr-x qatrack:www-data  qatrack/media/uploads   (2755 - created by Django, umask 022)
-rw-r--r-- qatrack:www-data  .../perm_test_….txt
```

**The stated premise checks out.** `/proc/<pid>/status` confirms *both* gunicorn and django-q2 run as `$USER` with `Umask: 0022`; `www-data` only runs nginx. setgid correctly propagates the `www-data` group onto directories Django creates later, so nginx can still read everything.

The `linux_upgrade_from_3.rst` block was validated separately by simulating a v3/Apache-era tree (everything `www-data`-owned) and confirming the new commands convert it to a working state.

Also green: `manage.py check` clean, the 6 new tests pass, `sphinx-build` produces no warnings on the changed `.rst` files, `make test_simple` → 1113 passed, all 11 CI checks green.

### Review comments

**1. `checks.py`: the "parent not writable" branch is unreachable in practice.** `qatrack/settings.py:765` runs `d.mkdir(parents=True, exist_ok=True)` at *import* time. So in the case that branch is written for — `media/uploads` missing and the parent unwritable — Django dies with a raw `PermissionError` traceback before the check framework ever runs. Verified:

```
File ".../qatrack/settings.py", line 765, in <module>
    d.mkdir(parents=True, exist_ok=True)
PermissionError: [Errno 13] Permission denied: '.../qatrack/media/uploads'
```

`test_parent_not_writable_for_missing_dir` only passes because `os.access` is mocked. Suggest either guarding the settings-level mkdir so the check can report it properly, or dropping that branch and its test. (The other branch works well — when the dirs exist but aren't writable, E001 fires with the new `$USER:www-data` hint, exactly as intended.)

**2. `chmod -R 775` is broader than needed.** On an upgrade this makes every existing media file world-readable *and* group-writable by the network-facing nginx user, and it sets `+x` on regular files (verified: `README.rst` became `-rwxrwxr-x`). `chmod -R u=rwX,g=rX,o=` with the same `g+s` works identically — I tested it end to end (upload OK, nginx 200) under both umask 022 and 027. Worth considering for a clinical app's media directory.

**3. `sudo usermod -a -G www-data $USER` is now vestigial.** I removed `qatrack` from the `www-data` group, restarted gunicorn and django-q2, and uploads plus nginx serving both still worked — setgid assigns the group regardless of membership, and nginx's traversal of `/home/$USER` is handled by `make nginx.conf`'s `usermod -a -G $(USER) www-data`. Corollary: **removing the `exec sg www-data newgrp …` line is safe** — nothing downstream in the same shell depends on the refreshed group.

**4.** `os.access(…, W_OK)` always returns True for root, so `sudo python manage.py check` reports "no issues" on a genuinely broken install (verified). Consider warning when `euid == 0`.

**5.** The two early `Error(...)`s ("The Media folder is not configured" / "does not exist") have no `id=`, unlike E001/E002, so they can't be filtered or silenced via `SILENCED_SYSTEM_CHECKS`.

**6.** `$USER` is literal text in the hint. Pasted into a root shell it would chown media to `root` and make things worse; `getpass.getuser()` would let the hint print an exact command.

**7.** This PR targets `master`, but the description says target branch `develop`. Both refs are at `f89ea39` right now so there's no diff impact, but flagging it against the "Reviewer verifies the target branch" checklist item.

**8.** Nits: `linux.rst` now has three consecutive blank lines after the new block; the upgrade doc's new block uses relative paths (`logs qatrack/media`) without a preceding `cd ~/web/qatrackplus` (the block *after* it re-cds, so it'd be safer to add).

### Separately: three pre-existing bugs in the install guide

Not caused by this PR, but I hit all three while following the guide verbatim, and the first two are hard stops:

- **Step order — `loaddata` is documented before `createcachetable` and fails:** `django.db.utils.ProgrammingError: Problem installing fixture '.../service_log/serviceeventstatus.json': relation "qatrack_cache_table" does not exist`. Moving `createcachetable` above `loaddata` fixes it (confirmed).
- **`gettext` is missing from the prerequisites.** `python manage.py compilemessages -l fr` dies with `CommandError: Can't find msgfmt`. `gettext` is not in `sudo apt install make build-essential python3-dev python3-tk curl` and is not present on a stock Noble image.
- **`compilemessages -l es` still exits non-zero after installing gettext.** `qatrack/locale/es/LC_MESSAGES/django.po` has 141 `msgfmt --check` errors — format-specification mismatches at lines 251 (`phrase`), 582, 934/958/988 (`fault`), and others. `fr` and `fr_CA` are clean.

Minor drift also worth a pass: the pg_hba path says `/etc/postgresql/18/main/pg_hba.conf` but Ubuntu 24.04 ships PostgreSQL **16** (as the guide's own intro states); the expected test output says `975 passed, 5 skipped` vs. an actual `1113 passed, 3 skipped`; and `showmigrations accounts` lists 4 migrations vs. 5 actual (`0005_v4_0_final`).

Happy to open a follow-up PR for those three if useful.

<details>
<summary>Test environment caveats</summary>

The VM's egress proxy blocked `dl.google.com` and `astral.sh`, so the Chrome/PDF-report step was skipped and `uv` was installed from PyPI rather than via the shell installer. The container also has no IPv6, so nginx could not bind until the default site was removed — which the guide instructs you to do anyway. Neither affected the permission behaviour under test.

</details>

**Comment by nsmela** [comment]
Discovered an annoying issue: the postgres version available is dependant on the Ubuntu version used. 16 for Ubuntu 24, 18 for Ubuntu 26 (which i use as a test server). I initially thought the changes to postgres 16 in the commands was an error, but it seems I had the wrong setup.

File: docs/install/backup.rst
Comment (crcrewso) [inline docs/install/backup.rst:1]: QATrack+ user is ambiguous, a quick reading could confuse in-app user versus OS level user. 
suggestion: `local OS user for QATrack+ services`

File: qatrack/qatrack_core/tests/test_checks.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    def test_media_folder_configured_and_writable(self):
+        import tempfile
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            media_path = Path(temp_dir)
+            (media_path / 'uploads').mkdir()
+            (media_path / 'uploads' / 'tmp').mkdir()
+            with override_settings(MEDIA_ROOT=str(media_path)):
+                errors = check_media_folder_permissions(None)
+                assert errors == []
```
[the remark this answers is not in the document]
Comment (crcrewso) [reply qatrack/qatrack_core/tests/test_checks.py:19]: @coderabbitai Suggested added cases should be listed as their own issue for maintainer workload balancing. Please create new issue with moderate level details

File: qatrack/qatrack_core/checks.py [hunk 2]
```diff
@@ -1,44 +1,70 @@
...
+
+    try:
+        import pwd
+
+        app_user = pwd.getpwuid(os.geteuid()).pw_name
+    except (ImportError, KeyError, AttributeError):
+        app_user = getpass.getuser()
+
+    if app_user == 'root':
+        app_user = '$USER'
```
[the remark this answers is not in the document]
Comment (crcrewso) [reply qatrack/qatrack_core/checks.py:47]: @coderabbitai tests should be executed with non-elevated user accounts. please provide comments in the developers guide to indicate this

**Review by crcrewso — COMMENTED** [review]
This branch is 90% of the way there! I never thought to separate out the errors and hints properly.

File: qatrack/qatrack_core/checks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nsmela) [reply qatrack/qatrack_core/checks.py]: Not taking the `SUDO_USER` suggestion — it would reintroduce a bug that was removed in 2b72034 and verified before removal.

`SUDO_USER` names whoever *invoked* sudo, not the account the process is running as. So the normal way an admin runs this check produces actively wrong advice:

```
$ sudo -u qatrack .venv/bin/python manage.py check
HINT: ... sudo chown -R siteadmin:www-data /home/qatrack/web/qatrackplus/qatrack/media ...
```

`siteadmin` is the admin's own account. Following that hint hands the media tree to the wrong user and re-breaks the install. In the same invocation, resolving the effective user returns `qatrack`, which is correct:

```
SUDO_USER= siteadmin | getpass.getuser()= qatrack
```

The current code resolves the effective user via `pwd.getpwuid(os.geteuid()).pw_name`, falling back to `getpass.getuser()` on platforms without `pwd`. The only case `SUDO_USER` would get right is a plain `sudo manage.py check` as root, and that is already covered both by the `root` fallback and by `qatrack.W001`.

That said, the concern behind the finding was valid and is now fixed, just differently. The root fallback was the literal string `$USER`, which is shell-expandable — a root operator pasting the hint would run `chown -R root:www-data` and make things worse. It is now `<qatrack-os-user>`, which cannot expand. Reachable via the `is_dir()` branch, where `os.access` does not short-circuit the check for root.

Also in the same commit: `MEDIA_ROOT` is now passed through `shlex.quote`, so a path containing spaces stays a single argument.

@crcrewso — the developer guide warning you asked for is in that commit too.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nsmela** [comment]
Not sure I want to pull any more threads on this one: the more effort I put into it, more issues that pop up.

**Comment by crcrewso** [comment]
> Not sure I want to pull any more threads on this one: the more effort I put into it, more issues that pop up.

I might have a different way of approaching this problem that we can investigate after the test suite is ready for review
